include/fmt/format.h

```diff
@@ -1798,22 +1818,76 @@ auto write_float(OutputIt out, const DecimalFP& fp,
   });
 }
 
+#ifdef __cpp_lib_bit_cast
+template <typename T, FMT_ENABLE_IF(is_fast_float<T>::value)>
+constexpr bool is_infinite_fast_float(T value) {
+  using floaty =
+      std::conditional_t<std::is_same<T, long double>::value, double, T>;
+  using uint = typename dragonbox::float_info<floaty>::carrier_uint;
+  constexpr auto significand_bits =
+      dragonbox::float_info<floaty>::significand_bits;
+  auto bits = std::bit_cast<uint>(value);
+  return (bits & exponent_mask<floaty>()) &&
+         !(bits & ((uint(1) << significand_bits) - 1));
+}
+
+template <typename T, FMT_ENABLE_IF(is_fast_float<T>::value)>
+constexpr bool is_finite_fast_float(T value) {
+  using floaty =
+      std::conditional_t<std::is_same<T, long double>::value, double, T>;
+  using uint = typename dragonbox::float_info<floaty>::carrier_uint;
+  auto bits = std::bit_cast<uint>(value);
+  return (bits & exponent_mask<floaty>()) != exponent_mask<floaty>();
+}
+#endif
+
 template <typename Char, typename OutputIt, typename T,
           FMT_ENABLE_IF(std::is_floating_point<T>::value)>
-auto write(OutputIt out, T value, basic_format_specs<Char> specs,
-           locale_ref loc = {}) -> OutputIt {
+FMT_CONSTEXPR20 auto write(OutputIt out, T value,
+                           basic_format_specs<Char> specs, locale_ref loc = {})
+    -> OutputIt {
   if (const_check(!is_supported_floating_point(value))) return out;
   float_specs fspecs = parse_float_type_spec(specs);
   fspecs.sign = specs.sign;
-  if (std::signbit(value)) {  // value < 0 is false for NaN so use signbit.
+  bool is_negative;
+#ifdef __cpp_if_constexpr
+  if (is_constant_evaluated()) {
+    // bit check below is valid only for "fast float" types
+    if constexpr (is_fast_float<T>::value) {
+      using floaty =
+          conditional_t<std::is_same<T, long double>::value, double, T>;
+      using uint = typename dragonbox::float_info<floaty>::carrier_uint;
+      auto bits = bit_cast<uint>(value);
+      is_negative = (bits & (uint(1) << (num_bits<uint>() - 1))) != 0;
```

Comment:
I think this can be simplified by bit casting into `char[sizeof(T)]` and checking the most-significant bit (using `is_big_endian` to determine it). Then `is_fast_float` can be replaced with `std::numeric_limits<T>::is_iec559`.

Also I suggest moving the sign check into a separate function, `fmt::detail::signbit`.

Reply:
> Also I suggest moving the sign check into a separate function, `fmt::detail::signbit`.

Sign check moved to a separate function. But, for example, this code:
https://github.com/fmtlib/fmt/blob/a842d3f84139cb897496c8f8751acac1ac4179d5/include/fmt/format.h#L1931-L1944
does bit casting of float to int twice now, so I think we shouldn't use `fmt::detail::float_signbit` here. And without using it here, there is only one usage of this function left, so I'm not sure that it's worth it at all.

Reply:
> I think this can be simplified by bit casting into `char[sizeof(T)]` and checking the most-significant bit (using `is_big_endian` to determine it). Then `is_fast_float` can be replaced with `std::numeric_limits<T>::is_iec559`.

`is_fast_float` is used here to omit `long double` as a type, since it can fulfill the requirements of IEC 559 (IEEE 754) standard, but even in this case, the most-significant bit can be not set - https://godbolt.org/z/ao1ab7eb9. As you [pointed out earlier in the discussion](https://github.com/fmtlib/fmt/issues/1895#issuecomment-864410482), the sign bit for most x86 is [79](https://en.wikipedia.org/wiki/Extended_precision#x86_extended_precision_format), not [127](https://en.wikipedia.org/wiki/Quadruple-precision_floating-point_format#IEEE_754_quadruple-precision_binary_floating-point_format:_binary128), but managing the sign bit number for each platform and each compiler seems.. let's say too tricky 😏.

But even with managing the sign bit number with some constants, the idea of bit casting a floating-point number into `char[]` instead of a single unsigned integer number looks more complicated than the current approach, which is already used in the library.

Also, `floaty` alias after `fast_float` check can be eliminated, since there are only IEEE 754 `float` and `double` left.

Reply:
> does bit casting of float to int twice now, so I think we shouldn't use fmt::detail::float_signbit here.

I think it's OK since cast should be optimized away.